odbc/src/api/data.rs

```diff
@@ -140,29 +142,89 @@ impl IntoCData for Bson {
+            Bson::Double(f) => Ok((
+                *f as u64,
+                if f.floor() != *f {
+                    Some(ODBCError::FractionalTruncation(f.to_string()))
+                } else {
+                    None
+                },
+            )),
+            Bson::String(s) => Ok((
+                u64::from_str(s)
+                    .map_err(|_| ODBCError::InvalidCharacterValue(s.clone(), UINT64))?,
+                None,
+            )),
+            Bson::Boolean(b) => Ok((u64::from(*b), None)),
+            Bson::Int32(i) if *i < 0i32 => Err(ODBCError::IntegralTruncation(i.to_string())),
+            Bson::Int32(i) => Ok((*i as u64, None)),
+            Bson::Int64(i) if *i < 0i64 => Err(ODBCError::IntegralTruncation(i.to_string())),
+            Bson::Int64(i) => Ok((*i as u64, None)),
+            // Note that this isn't perfect because there are some 64bit integer values that are
+            // not representable as doubles. There *could* be a specific value where we will get a
+            // different result here than if we had a conversion from Decimal128 to i64 directly.
+            // We should update this when the bson crate supports Decimal128 entirely.
+            Bson::Decimal128(_) => {
+                let out = self.to_f64()?;
+                if out < 0f64 {
+                    Err(ODBCError::IntegralTruncation(out.to_string()))
+                } else {
+                    Ok((
+                        out as u64,
+                        if out.floor() != out {
+                            Some(ODBCError::FractionalTruncation(out.to_string()))
+                        } else {
+                            None
+                        },
+                    ))
+                }
+            }
+            o => Err(ODBCError::RestrictedDataType(o.to_type_str(), UINT64)),
+        }
     }
 
     fn to_u32(&self) -> Result<(u32, Option<ODBCError>)> {
-        self.to_i32().map(|(i, w)| (i as u32, w))
+        self.to_u64().map_err(|e| match e {
+            ODBCError::RestrictedDataType(s, _) => ODBCError::RestrictedDataType(s, UINT32),
+            ODBCError::InvalidCharacterValue(s, _) => ODBCError::InvalidCharacterValue(s, UINT32),
+            _ => e,
+        })?;
+        match self {
+            Bson::Double(x) if *x > u32::MAX as f64 || *x < 0f64 => {
+                Err(ODBCError::IntegralTruncation(x.to_string()))
+            }
+            Bson::Int64(x) if *x > u32::MAX as i64 || *x < 0i64 => {
+                Err(ODBCError::IntegralTruncation(x.to_string()))
+            }
+            Bson::Decimal128(x) if self.to_f64()? > u32::MAX as f64 || self.to_f64()? < 0f64 => {
+                Err(ODBCError::IntegralTruncation(x.to_string()))
+            }
+            Bson::Int32(x) if *x < 0i32 => Err(ODBCError::IntegralTruncation(x.to_string())),
+            _ => self.to_u64().map(|(u, w)| (u as u32, w)),
```

Comment:
Maybe I'm missing something here, but we seem to be calling `to_u64` twice. The first time we just throw the result away.

Reply:
I was doing this to check for unsupported conversions. Thanks for calling this out. I'm refactoring it to `map_or_else` which will provide a much cleaner implementation!